VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp

```diff
@@ -0,0 +1,332 @@
+	return NewStaticMeshComponent;
+}
+
+void BlockUntilCookCompleted()
+{
+	FScopedSlowTask PRTGenerateCallsTasks(0, FText::FromString("Finishing previous Vitruvio cooking..."));
+	while (IsCooking.load())
+	{
+		FPlatformProcess::Sleep(0); // SwitchToThread
+		int32 CurrentNumGenerateCalls = VitruvioModule::Get().GetNumGenerateCalls();
+		PRTGenerateCallsTasks.EnterProgressFrame(0);
+	}
+}
+
+void BlockUntilGenerated()
+{
+	// Wait until all async generate calls to PRT are finished. We want to block the UI and show a modal progress bar.
+	int32 TotalGenerateCalls = VitruvioModule::Get().GetNumGenerateCalls();
+	FScopedSlowTask PRTGenerateCallsTasks(TotalGenerateCalls, FText::FromString("Generating models..."));
+	PRTGenerateCallsTasks.MakeDialog();
+	while (VitruvioModule::Get().IsGenerating() || VitruvioModule::Get().IsLoadingRpks())
+	{
+		FPlatformProcess::Sleep(0); // SwitchToThread
+		int32 CurrentNumGenerateCalls = VitruvioModule::Get().GetNumGenerateCalls();
+		PRTGenerateCallsTasks.EnterProgressFrame(TotalGenerateCalls - CurrentNumGenerateCalls);
+		TotalGenerateCalls = CurrentNumGenerateCalls;
+	}
+}
+
+UTexture2D* SaveTexture(UTexture2D* Original, const FString& Path, FTextureCache& TextureCache)
+{
+	if (TextureCache.Contains(Original))
+	{
+		return TextureCache[Original];
+	}
+
+	FString TexturePath = FPaths::Combine(Path, TEXT("Textures"), Original->GetName());
+	UPackage* TexturePackage = CreatePackage(*TexturePath);
+	UTexture2D* NewTexture = NewObject<UTexture2D>(TexturePackage, *Original->GetName(), RF_Public | RF_Standalone);
+
+	NewTexture->PlatformData = new FTexturePlatformData();
+	NewTexture->PlatformData->SizeX = Original->PlatformData->SizeX;
+	NewTexture->PlatformData->SizeY = Original->PlatformData->SizeY;
+	NewTexture->PlatformData->PixelFormat = Original->PlatformData->PixelFormat;
+	NewTexture->CompressionSettings = Original->CompressionSettings;
+	NewTexture->SRGB = Original->SRGB;
+
+	// Allocate first mipmap and upload the pixel data
+	FTexture2DMipMap* Mip = new FTexture2DMipMap();
+	FTexture2DMipMap& OriginalMip = Original->PlatformData->Mips[0];
+
+	NewTexture->PlatformData->Mips.Add(Mip);
+
+	Mip->SizeX = OriginalMip.SizeX;
+	Mip->SizeY = OriginalMip.SizeY;
+
+	const uint8* SourcePixels = static_cast<const uint8*>(OriginalMip.BulkData.LockReadOnly());
+	Mip->BulkData.Lock(LOCK_READ_WRITE);
+
+	void* TextureData = Mip->BulkData.Realloc(CalculateImageBytes(Mip->SizeX, Mip->SizeY, 0, NewTexture->PlatformData->PixelFormat));
+	FMemory::Memcpy(TextureData, SourcePixels, OriginalMip.BulkData.GetBulkDataSize());
+	Mip->BulkData.Unlock();
+
+	NewTexture->Source.Init(Original->PlatformData->SizeX, Original->PlatformData->SizeY, 1, 1, ETextureSourceFormat::TSF_BGRA8, SourcePixels);
+	OriginalMip.BulkData.Unlock();
+
+	NewTexture->PostEditChange();
+	TexturePackage->MarkPackageDirty();
+	FAssetRegistryModule::AssetCreated(NewTexture);
+
+	TextureCache.Add(Original, NewTexture);
+
+	return NewTexture;
+}

```

Comment:
Would it be possible to replace this code with a generic DuplicateAsset call (e.g. SequenceRecorderUtils.h::DuplicateAsset)?